### PR TITLE
[SMTChecker] Refactor CHC predicates

### LINKED ISSUE #9414: [SMTChecker] Refactor CHC predicates
The predicates in CHC should be refactored into their own class which would contain the `Sort,` `SymbolicVariable`, `ASTNode`, potential concrete values for its arguments and how to format them as counterexample.

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes #9414 

Creates a new class `Predicate` that wraps `SymbolicFunctionVariable`, used as predicates in the CHC engine.
The predicates are owned by the static map `m_predicates`, as suggested by @chriseth , and the maps in CHC only use pointers to those predicates.
Furthermore, this PR also refactors the code that constructs counterexamples, mostly moving code into the new Predicate class, making the code more robust.

This PR is more easily reviewed by looking at each commit sequentially.

### CODE REVIEW

File: libsolidity/formal/CHC.cpp [hunk 1]
```diff
@@ -923,8 +918,10 @@ void CHC::defineInterfacesAndSummaries(SourceUnit const& _source)
 						auto state1 = stateVariablesAtIndex(1, *base);
 						auto state2 = stateVariablesAtIndex(2, *base);
 
-						auto nondetPre = iface(state0 + state1);
-						auto nondetPost = iface(state0 + state2);
+						auto const* iface = m_nondetInterfaces.at(base);
+						auto state0 = stateVariablesAtIndex(0, *base);
+						auto nondetPre = (*iface)(state0 + state1);
+						auto nondetPost = (*iface)(state0 + state2);
```
Comment (mijovic) [inline libsolidity/formal/CHC.cpp:909]: Personally, for me (probably anyone else looking this for the first time) it is too hard to follow all the types here when over-using auto. Anyway, I am fine with it.

File: libsolidity/formal/CHC.cpp [hunk 2]
```diff
@@ -994,37 +991,36 @@ smtutil::Expression CHC::summary(FunctionDefinition const& _function)
...
 	auto block = createSymbolicBlock(sort(_node),
 		"block_" +
-		uniquePrefix() +
-		"_" +
-		_prefix +
-		predicateName(_node));
+			uniquePrefix() +
+			"_" +
+			_prefix +
+			predicateName(_node),
```
Comment (mijovic) [inline libsolidity/formal/CHC.cpp]: Maybe can be done in one line, updating this one and deleting `"block_" +`
```suggestion
		"block_" + uniquePrefix() + "_" + _prefix + predicateName(_node),
```
But, I am fine with multi-line version if you prefer it more for readability.

File: libsolidity/formal/CHC.cpp [hunk 3]
```diff
@@ -994,37 +991,36 @@ smtutil::Expression CHC::summary(FunctionDefinition const& _function)
...
+Predicate const* CHC::createSummaryBlock(FunctionDefinition const& _function, ContractDefinition const& _contract)
 {
 	auto block = createSymbolicBlock(summarySort(_function, _contract),
 		"summary_" +
-		uniquePrefix() +
-		"_" +
-		predicateName(&_function, &_contract));
+			uniquePrefix() +
+			"_" +
+			predicateName(&_function, &_contract),
```
Comment (mijovic) [inline libsolidity/formal/CHC.cpp]: Same think about one vs multi-line

File: libsolidity/formal/Predicate.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	string _name,
+	EncodingContext& _context,
+	ASTNode const* _node
+):
+	m_predicate(_sort, move(_name), _context),
+	m_node(_node)
+{
+}
+
+Predicate const* Predicate::predicate(std::string const& _name)
```
Comment (mijovic) [inline libsolidity/formal/Predicate.cpp]:
```suggestion
Predicate const* Predicate::predicate(string const& _name)
```

File: libsolidity/formal/Predicate.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+{
+	return &m_predicates.at(_name);
+}
+
+void Predicate::reset()
+{
+	m_predicates.clear();
+}
+
+smtutil::Expression Predicate::operator()(std::vector<smtutil::Expression> const& _args) const
```
Comment (mijovic) [inline libsolidity/formal/Predicate.cpp]:
```suggestion
smtutil::Expression Predicate::operator()(vector<smtutil::Expression> const& _args) const
```

File: libsolidity/formal/Predicate.h [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	static Predicate const* create(
+		smtutil::SortPointer _sort,
+		std::string _name,
+		smt::EncodingContext& _context,
+		ASTNode const* _node = nullptr
+	);
+
+	static Predicate const* predicate(std::string const& _name);
+
+	static void reset();
```
Comment (mijovic) [inline libsolidity/formal/Predicate.h:59]: I know it should be obvious, but doesn't hurt adding docstring that says reset is clearing all predicates

**Review by mijovic — COMMENTED** [review]
This is review of first commit. Will submit one review per commit, as I think it is easier to review.

Looks good, only some small comments about style.

File: test/libsolidity/smtCheckerTests/modifiers/modifier_code_after_placeholder.sol [hunk 7]
```diff
@@ -21,5 +21,5 @@ contract C
 	}
 }
 // ----
-// Warning 4984: (203-208): Overflow (resulting value larger than 2**256 - 1) happens here
+// Warning 4984: (203-208): Overflow (resulting value larger than 2**256 - 1) happens here.
```
Comment (mijovic) [inline test/libsolidity/smtCheckerTests/modifiers/modifier_code_after_placeholder.sol:24]: I am curious about how this happens, when it has `.` and when not?

**Review by mijovic — COMMENTED** [review]
Includes a lot of refactoring and moving things. Generally looks good, I left few small style suggestions

File: test/libsolidity/smtCheckerTests/modifiers/modifier_code_after_placeholder.sol (hunk 7 above)
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/modifiers/modifier_code_after_placeholder.sol:24]: When the SMT query is satisfiable, usually meaning "there is an overflow, assertion failure, etc", we want to get a counterexample for the failed property. The problem with doing that in CHC is that the solver applies a lot of preprocessing that make retrieving counterexamples impossible. So we do a second run (same query) but with solver preprocessing disabled. The problem is now much harder for the solver, so it might timeout or answer `unknown`. If the solver still says SAT, we print a counterexample, so the `.` is not there. If the problem without preprocessing was indeed too hard and the solver couldn't solve it, we still know that the problem is SAT, so we can report "overflow etc", but we don't have a counterexample, so the message ends there with `.`.

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/modifiers/modifier_code_after_placeholder.sol:24]: Ah, and the reason this changed here in this PR is that the encoding changed a bit, and this sometimes triggers different solving behaviors in the solver.

Comment (mijovic) [reply test/libsolidity/smtCheckerTests/modifiers/modifier_code_after_placeholder.sol:24]: Thanks for clarification

File: libsolidity/formal/CHC.cpp (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp:909]: I agree with you that at a first glance that might be the case, but when you're used to this file you know that these are Predicates applications (smtutil::Expression). In any case, I'm also fine with either and will change if you think it makes sense.

**Comment by leonardoalt** [comment]
Updated review comments

File: libsolidity/formal/CHC.cpp (hunk 1 above)
Comment (mijovic) [reply libsolidity/formal/CHC.cpp:909]: I am fine with leaving auto

**Review by mijovic — DISMISSED** [review]
LGTM.
Maybe only squashing last commit would be nice to do.

**Comment by leonardoalt** [comment]
@mijovic @ekpyron updated

File: libsolidity/formal/Predicate.h [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+public:
+	static Predicate const* create(
+		smtutil::SortPointer _sort,
+		std::string _name,
+		smt::EncodingContext& _context,
+		ASTNode const* _node = nullptr
+	);
+
+	/// Predicate should be movable.
+	Predicate(Predicate&&) = default;
```
Comment (leonardoalt) [inline libsolidity/formal/Predicate.h]: Wouldn't compile without this

Comment (ekpyron) [inline libsolidity/formal/Predicate.h]: If you have this, maybe you should also have ``Predicate& operator=(Predicate&&) = default;``? But fine without it, if things work without it...

The main thing I'm wondering is:
How can they be movable in the first place, if you have a static ``create`` function, which is supposed to keep them inside a static map :-)?
To me it seems that if you have that, then this cannot even be movable, since you don't want to ever move it out of this map...

Are they only movable, because they have to be move-constructible for the map ``emplace``?
If so, then just construct in-place instead, i.e. use ``m_predicates.emplace(std::piecewise_construct, std::forward_as_tuple(functorName), std::forward_as_tuple(_sort, _name, _context, _node));`` in the implementation of ``create`` instead - then you should be able to delete the move constructor (if not then there's something wrong somewhere, right?)

Comment (ekpyron) [reply libsolidity/formal/Predicate.h]: Haha, yeah, just commented about this - I think this should be deleted as well and you should change the map emplace to an in-place emplace, then it should still compile without being even movable.

Comment (ekpyron) [reply libsolidity/formal/Predicate.h]: Ah, hm... you need to construct it first to get the name under which to store it in the map the way things are built right now? Hm... that's not so nice... let me see...

File: libsolidity/formal/Predicate.cpp [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+Predicate::Predicate(
+	SortPointer _sort,
+	string _name,
+	EncodingContext& _context,
+	ASTNode const* _node
+):
+	m_predicate(_sort, move(_name), _context),
+	m_node(_node)
+{
+}
```
Comment (ekpyron) [inline libsolidity/formal/Predicate.cpp]: So something like this maybe :-)? Together with deleting the move constructor (and adjusting the constructor in the header)?
```suggestion
Predicate const* Predicate::create(
	SortPointer _sort,
	string _name,
	EncodingContext& _context,
	ASTNode const* _node
)
{
	smt::SymbolicFunctionVariable predicate{_sort, move(_name), _context};
	string functorName = predicate.name;
	solAssert(!m_predicates.count(functorName), "");
	return &m_predicates.emplace(std::piecewise_construct, std::forward_as_tuple(functorName), std::forward_as_tuple(move(predicate), _node)).first->second;
}

Predicate::Predicate(
	smt::SymbolicFunctionVariable&& _predicate,
	ASTNode const* _node
):
	m_predicate(std::move(_predicate)),
	m_node(_node)
{
}
```

File: libsolidity/formal/Predicate.h (hunk 8 above)
Comment (leonardoalt) [reply libsolidity/formal/Predicate.h]: Yea I wanted to add the move assign operator as well, but then it started complaining about `m_context` being a reference, so I left it out. Wouldn't be dangerous since the copy version is deleted anyway.

That does make sense.
I think in the first version, where the ownership was different, CHC also owned some, so I moved some predicates around.
But now I can't find any other move case. I'll try your suggestion.

Comment (leonardoalt) [reply libsolidity/formal/Predicate.h]: Yes exactly

File: libsolidity/formal/Predicate.cpp (hunk 9 above)
Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp]: lol let me try

File: libsolidity/formal/Predicate.h (hunk 8 above)
Comment (ekpyron) [reply libsolidity/formal/Predicate.h]: I just put up a lengthier suggestion - since the constructor is private you can just as well construct the ``SymbolicFunctionVariable`` in ``create`` instead and just move it down the constructor... I think that should work.

Comment (ekpyron) [reply libsolidity/formal/Predicate.h]: An alternative would be to make the static map actually a ``set`` that has a transparent comparison operator that compares according to the name (and being transparent allows lookups by name) :-D... but that might be extreme overkill :-) (wanting the move constructor deleted may already qualify as slight overkill ;-)).

File: libsolidity/formal/Predicate.cpp (hunk 9 above)
Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp]: This worked, the only annoying thing is that now the actual constructor needs to be public, and I'm not sure there would be a way to guarantee that it's owned by the map. Thinking about it

Comment (ekpyron) [reply libsolidity/formal/Predicate.cpp]: Ah yeah, hadn't thought about that... that's a bit annoying, that's true...

Comment (ekpyron) [reply libsolidity/formal/Predicate.cpp]: There's no *good* solution to this... the go-to way if you *really* want it, is something like https://gist.github.com/valmat/69c67f94daf75452e2ae987aa67d55be , but that's kind of horrible...

Comment (ekpyron) [reply libsolidity/formal/Predicate.cpp]: Not sure this is worth too much time - it'd be nicer if this ``Predicate`` *wasn't* movable, but it's actually just fine if it is, especially since the ``predicate(std::string const& _name);`` returns a ``const`` ref only anyways, so there won't be any public way to move anything out of the map anyways...

Comment (ekpyron) [reply libsolidity/formal/Predicate.cpp]: The fact that this *would* work with a public constructor at least tells us that these *aren't* in fact moved around anywhere, so I'd be fine with moving back to (resp. staying with) the ``default``ed move constructor version.

File: libsolidity/formal/CHC.cpp [hunk 10]
```diff
@@ -681,7 +673,9 @@ void CHC::resetSourceAnalysis()
 	m_errorIds.clear();
 	m_callGraph.clear();
 	m_summaries.clear();
-	m_symbolFunction.clear();
+	m_interfaces.clear();
+	m_nondetInterfaces.clear();
+	Predicate::reset();
```
Comment (ekpyron) [inline libsolidity/formal/CHC.cpp:678]: By the way: I'd say it's fine for now, but in general, it'd be nicer to have explicit ownership here (i.e. make the map a member of ``CHC`` like all the others above and just clear it here, making ``Predicate::create`` a member of ``CHC`` instead... in which case a public constructor in ``Predicate``, but it being non-movable makes more sense...)... but I don't think it's worth changing this now (and you probably had reasons not to go that way anyways).

File: libsolidity/formal/CHC.cpp [hunk 11]
```diff
@@ -1128,13 +1106,13 @@ string CHC::predicateName(ASTNode const* _node, ContractDefinition const* _contr
...
 
-smtutil::Expression CHC::predicate(smt::SymbolicFunctionVariable const& _block)
+smtutil::Expression CHC::predicate(Predicate const& _block)
 {
 	return _block(currentBlockVariables());
 }
 
 smtutil::Expression CHC::predicate(
-	smt::SymbolicFunctionVariable const& _block,
+	Predicate const& _block,
```
Comment (ekpyron) [inline libsolidity/formal/CHC.cpp:1115]: Actually I'm not quite sure if these two functions help readability that much... to me it seems it'd be easier to see what happens, if you just inline every call to these two, won't it?
But also nothing that needs to be done now...

**Review by ekpyron — DISMISSED** [review]
Yeah, I think we can improve further on top of this, but this is "good enough" for now and definitely a step in the right direction - so I'm approving, even though I haven't reviewed in every detail (which I hope @mijovic did earlier ;-)).

**Review by mijovic — DISMISSED** [review]

File: libsolidity/formal/CHC.cpp (hunk 10 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp:678]: The reason to not go that way is that I kinda wanted to decouple CHC from Predicates. But also wouldn't mind it the way you described.

File: libsolidity/formal/Predicate.cpp (hunk 9 above)
Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp]: I think it's fine the way it is now, just need to keep in mind to never actually use that public constructor except in very local/special cases ;)

File: libsolidity/formal/CHC.cpp (hunk 11 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp:1115]: True

**Review by ekpyron — APPROVED** [review]

File: libsolidity/formal/CHC.cpp (hunk 10 above)
Comment (ekpyron) [reply libsolidity/formal/CHC.cpp:678]: The thing is: static members with resets like this are always a bit hacky IMO... and at the latest if we ever want the compiler to be re-entrant (i.e. we allow multiple instances of the compiler to live in a process) all of them have to go... maybe we'll never do that, though :-)... but yeah, fine to keep it for now.
